README
This program is similar to ping(1), but tests connectivity at the
link layer (layer-2) instead of the network layer (layer-3).

HPUX has linkloop support for ages, and lately I googled a little
and found that link-layer connectivity loopback testing is part
of the IEEE 802.2LLC standard.

So I went and written:
	- linkloop for GNU/Linux.
	- linkloop_reply as a user space reply program. Used for
	  testing against GNU/Linux until the Linux kernel has native
	  support for this protocol (in my TODO :-)
	- lanscan is a small utility to show local interfaces
	  with their mac addresses. Yes I know 'ip link' gives
	  the same information, I just wanted to play with some
	  ioctls :-)

To get me started I grabbed some examples from spak
	http://www.xenos.net/software/spak/
(Which wasn't compiling cleanly on my machine BTW).

It was tested against HPUX-11.0 and GNU/Linux (when running linkloop_reply)
both ways.

My linkloop accepts MAC address in:
	- HPUX notation (0x......)
	- Colon separated (00:BC:.....)
	- A name from /etc/ethers


NEWS
* Changes from 0.0.1
  - Updated man page documenting all options
  - Better validation and reporting of received packets


common.c
/*
 * Copyright (C) 2004, 2005, 2006, 2007 Oron Peled
 *
 * Some code snippets adapted from spak
 * (http://www.xenos.net/software/spak/)
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

static const char rcsid[] = "$Id: common.c,v 1.1 2007/01/20 22:29:39 oron Exp $";

#include <netinet/in.h>		/* for htons(3)			*/

#include <net/if.h>		/* for IFNAMSIZ, IFHWADDRLEN */
#include <sys/ioctl.h>		/* for SIOCGIFHWADDR		*/
#include <stdio.h>
#include <string.h>     	/* for memcmp(3) memset(3)	*/
#include <stdlib.h>		/* for exit(3)			*/
#include <errno.h>
#include <assert.h>
#include "linkloop.h"
#if HAVE_ETHER_HOSTTON
#include <netinet/ether.h>	/* for ether_hostton(3)		*/
#endif

/* These are used to define different data structs... */
#if IFHWADDRLEN != ETH_ALEN
#error "SOMETHING IS VERY FISHY: IFHWADDRLEN != ETH_ALEN"
#endif

int debug_flag = 0;

void dump_packet(const struct llc_packet *pack)
{
	int i;
	uint8_t *p = (uint8_t *)pack;
	size_t len = ntohs(pack->eth_hdr.ether_type);

	printf("PACKET DUMP: data size=%d (0x%x)", len, len);
	for(i = 0; i < len; i++, p++) {
		if((i % 16) == 0)
			printf("\n%04x\t", i);
		printf("%02x ", (unsigned)*p);
	}
	printf("\nEND PACKET DUMP\n");
}

char *mac2str(const uint8_t s[])
{
	static char buf[3 * IFHWADDRLEN + 1];

	sprintf (buf, "%02X:%02X:%02X:%02X:%02X:%02X",
			s[0], s[1], s[2], s[3], s[4], s[5]);
	return buf;
}

#if 0
/* FIXME: use ether_ntoa() instead of mac2str() */
char *ether_ntoa(const struct ether_addr *addr);
#endif

int parse_address(uint8_t mac[], const char *str)
{
	unsigned a, b, c, d, e, f;
	struct ether_addr ea;

	if(sscanf(str,"%x:%x:%x:%x:%x:%x", &a, &b, &c, &d, &e, &f) == 6) {
		/* A colon separated notation */
		mac[0] = (unsigned char) a;
		mac[1] = (unsigned char) b;
		mac[2] = (unsigned char) c;
		mac[3] = (unsigned char) d;
		mac[4] = (unsigned char) e;
		mac[5] = (unsigned char) f;
	} else if(sscanf(str,
			"0x%02x%02x%02x%02x%02x%02x", &a, &b, &c, &d, &e, &f) == 6) {
		/* Hexadecimal notation (like HPUX) */
		mac[0] = (unsigned char) a;
		mac[1] = (unsigned char) b;
		mac[2] = (unsigned char) c;
		mac[3] = (unsigned char) d;
		mac[4] = (unsigned char) e;
		mac[5] = (unsigned char) f;
#if HAVE_ETHER_HOSTTON
	} else if(ether_hostton(str, &ea) == 0) {
		/* A name from /etc/ethers */
		memcpy(mac, ea.ether_addr_octet, 8);
#endif
	} else
		return 0;
	return 1;
}

void get_hwaddr(int sock, const char name[], uint8_t mac[])
{
	struct ifreq ifr;

	/* Make sure we have space for null terminator */
	strncpy(ifr.ifr_name, name, sizeof(ifr.ifr_name) - 1);
	if(ioctl(sock, SIOCGIFHWADDR, &ifr) < 0) {
		perror("ioctl(SIOCGIFHWADDR)");
		exit(1);
	}
	memcpy(mac, ifr.ifr_hwaddr.sa_data, IFHWADDRLEN);
}

void mk_test_packet(struct llc_packet *pack, const uint8_t src[], const uint8_t dst[], size_t len, int response)
{
	int i;

	if(len > PACK_SIZE) {		/* 0x05DC == 1500 */
		fprintf(stderr, "packet too long (%d bytes)\n", len);
		exit(1);
	}
	if(debug_flag)
		printf("Create test packet (%d bytes)\n", len);
	memcpy(pack->eth_hdr.ether_dhost, dst, sizeof(pack->eth_hdr.ether_dhost));
	memcpy(pack->eth_hdr.ether_shost, src, sizeof(pack->eth_hdr.ether_shost));
	pack->eth_hdr.ether_type = htons(len);
	pack->llc.dsap = (response) ? 0x80 : 0x00;
	pack->llc.ssap = (response) ? 0x01 : 0x80;	/* XNS? */
	pack->llc.ctrl = TEST_CMD;			/* TEST */
	/* Payload */
	// FIXME: ? len -= sizeof(struct llc) + sizeof(struct ether_header);
	for(i = 0; i < len; i++)
		pack->data[i] = i;
}

void send_packet(int sock, const char iface[], struct llc_packet *pack)
{
	struct sockaddr sa;
	int ret;

	memset((char *)&sa, 0, sizeof(sa));
	sa.sa_family = AF_INET;
	strncpy(sa.sa_data, iface, sizeof(sa.sa_data) - 1);
	/* Send the packet */
	ret = sendto(sock, pack, sizeof(*pack), 0, (struct sockaddr *)&sa, sizeof(sa));
	if(ret == -1) {
		perror("sendto");
		exit(1);
	}
	if(ret != sizeof(*pack))
		fprintf(stderr, "Warning: Incomplete packet sent\n");
	if(debug_flag)
		printf("sent TEST packet to %s\n", mac2str(pack->eth_hdr.ether_dhost));
}

int recv_packet(int sock, struct llc_packet *pack)
{
	struct sockaddr sa;
	socklen_t len;
	int ret;

	len = sizeof(sa);
	ret = recvfrom(sock, pack, sizeof(*pack), 0, (struct sockaddr *)&sa, &len);
	if(ret == -1) {
		if(errno == EINTR)	/* We have a timeout */
			return 0;
		perror("recvfrom");
		exit(1);
	}
	if((pack->llc.ctrl & TEST_CMD) != TEST_CMD) {
		fprintf(stderr, "got unexpected packet\n");
		dump_packet(pack);
		exit(1);
	}
	ret = ntohs(pack->eth_hdr.ether_type);
	if(debug_flag)
		printf("received TEST packet (%d bytes) from %s\n", ret, mac2str(pack->eth_hdr.ether_shost));
	return ret;
}



lanscan.c
/*
 * Copyright (C) 2004, 2005, 2006, 2007 Oron Peled
 *
 * Some code snippets adapted from spak
 * (http://www.xenos.net/software/spak/)
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <net/if.h>
#include "config.h"
#include "linkloop.h"

static const char rcsid[] = "$Id: lanscan.c,v 1.1 2007/01/20 22:29:39 oron Exp $";

#define	MAX_IFACES	20
#define	OPTSTRING	"V"
const char *program = NULL;
const char *arg;

void usage()
{
	fprintf(stderr, "Usage: %s [option...]\n"
		"\t-V		Print program version\n"
		, program
	);
	exit(1);
}

void handle_options(int argc, char * const argv[])
{
	int c;

	program = argv[0];
	while((c = getopt(argc, argv, OPTSTRING)) != -1)
		switch(c) {
		case 'V':
			printf("%s version %s\n%s", program, VERSION, COPYRIGHT_NOTICE);
			exit(0);
		default:
			fprintf(stderr, "%s: unknown option code 0x%x\n", program, c);
		case '?':
			usage();
		}
	if(optind != argc) {
		fprintf(stderr, "%s: Extra paramaters\n", program);
		usage();
	}
	arg = argv[optind];
}

#define	IFF(x)	{ IFF_ ## x, #x }

struct if_flag {
	short int flag;
	char *flag_name;
} if_flags[] = {
	IFF(ALLMULTI),
	IFF(BROADCAST),
	IFF(DEBUG),
	IFF(LOOPBACK),
	IFF(NOARP),
	IFF(NOTRAILERS),
	IFF(POINTOPOINT),
	IFF(PROMISC),
	IFF(RUNNING),
	IFF(UP),
#ifdef	LINUX
	IFF(AUTOMEDIA),
	IFF(MASTER),
	IFF(SLAVE),
	IFF(PORTSEL),
#endif
#ifdef	HPUX
	IFF(ALT_SR8025),
	IFF(AR_SR8025),
	IFF(CANTCHANGE),
	IFF(CKO),
	IFF(CKO_ETC),
	IFF(LFP),
	IFF(LNP),
	IFF(LOCALSUBNETS),
	IFF(MULTI_BCAST),
	IFF(MULTICAST),
	IFF(NOACC),
	IFF(NOSR8025),
	IFF(OACTIVE),
	IFF(UNNUMBERED),
#endif
};
#undef	IFF

short int get_ifflags(int sock, const char name[])
{
	struct ifreq ifr;

	strncpy(ifr.ifr_name, name, IF_NAMESIZE - 1);
	if (ioctl(sock, SIOCGIFFLAGS, &ifr) < 0) {
		perror("ioctl(SIOCGIFFLAGS)");
		exit(1);
	}
	return ifr.ifr_flags;
}

int main(int argc, char * const argv[])
{
	struct ifconf ifc;
	struct ifreq ifr_x[MAX_IFACES];
	int sock, err;
	uint8_t myMAC[IFHWADDRLEN];

	handle_options(argc, argv);
	if ((sock = socket(PF_PACKET, SOCK_PACKET, 0)) < 0) {
		perror("socket");
		exit(1);
	}

	ifc.ifc_len = MAX_IFACES * sizeof(struct ifreq);
	ifc.ifc_req = ifr_x;
	if((err = ioctl(sock, SIOCGIFCONF, &ifc)) < 0) {
		perror("ioctl");
		exit(1);
	}
	printf("retrieved info for %i interface(s)\n", ifc.ifc_len / sizeof(struct ifreq));
	for (err = 0; err < ifc.ifc_len / sizeof(struct ifreq); err++) {
		int i;
		short int flags = get_ifflags(sock, ifr_x[err].ifr_name);
		get_hwaddr(sock, ifr_x[err].ifr_name, myMAC);
		printf("%s\t%s\t<", ifr_x[err].ifr_name, mac2str(myMAC));
		for(i = 0; i < sizeof(if_flags)/sizeof(if_flags[0]); i++)
			if(flags & if_flags[i].flag)
				printf(" %s", if_flags[i].flag_name);
		printf(" >\n");
	}
	return EXIT_SUCCESS;
}


linkloop.h
/*
 * Copyright (C) 2004, 2005, 2006, 2007 Oron Peled
 *
 * Some code snippets adapted from spak
 * (http://www.xenos.net/software/spak/)
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

#ifndef	LINKLOOP_H
#define	LINKLOOP_H

#include "config.h"
#include <inttypes.h>
#include <net/if_arp.h>
#include <net/ethernet.h>

#define	COPYRIGHT_NOTICE	\
		"\nCopyright (C) 2003, 2005, 2006 Oron Peled\n" \
		"This is free software.  You may redistribute copies of it under the terms of\n" \
		"the GNU General Public License <http://www.gnu.org/licenses/gpl.html>.\n" \
		"There is NO WARRANTY, to the extent permitted by law.\n\n"

#define	PACKED		__attribute__((packed))
#define	PACK_SIZE	1514	/* This is what HPUX sends and responds to */
#define	PAYLOAD_SIZE	(PACK_SIZE - sizeof(struct llc) - sizeof(struct ether_header))

/* Describe an 802.2LLC packet */
struct llc {
	unsigned char dsap;
	unsigned char ssap;
	unsigned char ctrl;
} PACKED;

struct llc_packet {
	struct ether_header eth_hdr;
	struct llc	llc;
	/* Some test data */
	unsigned char data[PAYLOAD_SIZE];
} PACKED;

#define	TEST_CMD	0xE3	/* From 802.2LLC (HPUX sends 0xF3) */

void dump_packet(const struct llc_packet *pack);
char *mac2str(const uint8_t *s);
int parse_address(uint8_t mac[], const char *str);
void get_hwaddr(int sock, const char name[], uint8_t mac[]);
void mk_test_packet(struct llc_packet *pack, const uint8_t src[], const uint8_t dst[], size_t len, int response);
void send_packet(int sock, const char iface[], struct llc_packet *pack);
int recv_packet(int sock, struct llc_packet *pack);

#endif	/* LINKLOOP_H */


linkloop_reply.c
/*
 * Copyright (C) 2004, 2005, 2006, 2007 Oron Peled
 *
 * Some code snippets adapted from spak
 * (http://www.xenos.net/software/spak/)
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <net/if_arp.h>
#include <netinet/if_ether.h>
#include <net/if.h>
#include "config.h"
#include "linkloop.h"

static const char rcsid[] = "$Id: linkloop_reply.c,v 1.1 2007/01/20 22:29:39 oron Exp $";
static const char	*program = NULL;
static const char	*arg;
static char		*iface = "eth0";
static uint8_t		mac_src[IFHWADDRLEN];
static uint8_t		mac_dst[IFHWADDRLEN];
extern int		debug_flag;

#define	OPTSTRING	"di:V"

void usage()
{
	fprintf(stderr, "Usage: %s [option...]\n"
		"\t-d		Debug\n"
		"\t-i<iface>	Network interface\n"
		"\t-V		Print program version\n"
		, program
	);
	exit(1);
}

void handle_options(int argc, char * const argv[])
{
	int c;

	program = argv[0];
	while((c = getopt(argc, argv, OPTSTRING)) != -1)
		switch(c) {
		case 'd':
			debug_flag = 1;
			break;
		case 'i':
			iface = optarg;
			break;
		case 'V':
			printf("%s version %s\n%s", program, VERSION, COPYRIGHT_NOTICE);
			exit(0);
		default:
			fprintf(stderr, "%s: unknown option code 0x%x\n", program, c);
		case '?':
			usage();
		}
	if(debug_flag)
		fprintf(stderr, "%s: interface=%s\n", program, iface);
	if(optind != argc) {
		fprintf(stderr, "%s: Extra paramaters\n", program);
		usage();
	}
	arg = argv[optind];
}

int main(int argc, char *argv[])
{
	int sock;
	size_t len;
	struct llc_packet spack;
	struct llc_packet rpack;

	handle_options(argc, argv);

	/* Open a socket */
	if ((sock = socket(PF_PACKET, SOCK_PACKET, htons(ETH_P_802_2))) == -1) {
		perror("socket");
		return 1;
	}
	get_hwaddr(sock, iface, mac_src);
	printf ("Waiting for Test packets on %s (HW addr %s)\n", iface, mac2str(mac_src));
	do {
		len = recv_packet(sock, &rpack);
		memcpy(mac_dst, rpack.eth_hdr.ether_shost, IFHWADDRLEN);
		printf ("Got packet from: %s\n", mac2str(mac_dst));
		mk_test_packet(&spack, mac_src, mac_dst, len, 1);
		send_packet(sock, iface, &spack);
	} while(1);
	return EXIT_SUCCESS;
}


linkloop.c
/*
 * Copyright (C) 2004, 2005, 2006, 2007 Oron Peled
 *
 * linkloop: Loopback testing for layer-2 connectivity
 *           via 802.2LLC TEST message.
 *
 * Tested against HPUX-11.0
 *
 * Some code snippets adapted from spak
 * (http://www.xenos.net/software/spak/)
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

static const char rcsid[] = "$Id: linkloop.c,v 1.1 2007/01/20 22:29:39 oron Exp $";

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <getopt.h>
#include <unistd.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <net/if.h>	/* for IFNAMSIZ, IFHWADDRLEN */
#include <arpa/inet.h>
#include "config.h"
#include "linkloop.h"

/* Statistic counters */
static unsigned total_sent = 0;
static unsigned total_good = 0;
static unsigned total_timeout = 0;
static unsigned total_bad = 0;

/* Configurable options */
size_t	pack_size = ETHERMTU;	/* 0x05DC == 1500 */
int	timeout = 2;
int	retries = 1;
char	*iface = "eth0";

extern int	debug_flag;

#define	OPTSTRING	"di:t:n:s:V"
const char *program = NULL;
const char *arg;

void usage()
{
	fprintf(stderr, "Usage: %s [option...] mac_addr\n"
		"\t-d		Debug\n"
		"\t-i<iface>	Network interface\n"
		"\t-t<timeout>	Timeout between packets\n"
		"\t-n<retries>	Number of retries\n"
		"\t-s<size>	Packet size in bytes\n"
		"\t-V		Print program version\n"
		, program
	);
	exit(1);
}

void handle_options(int argc, char * const argv[])
{
	int c;

	program = argv[0];
	while((c = getopt(argc, argv, OPTSTRING)) != -1)
		switch(c) {
		case 'd':
			debug_flag = 1;
			break;
		case 'i':
			iface = optarg;
			break;
		case 't':
			timeout = strtol(optarg, NULL, 0);
			break;
		case 'n':
			retries = strtol(optarg, NULL, 0);
			break;
		case 's':
			pack_size = strtol(optarg, NULL, 0);
			if(pack_size > ETHERMTU) {
				fprintf(stderr, "%s: Illegal size specified (%d) > ETHERMTU (%d)\n",
					program, pack_size, ETHERMTU);
				exit(1);
			}
			break;
		case 'V':
			printf("%s version %s\n%s", program, VERSION, COPYRIGHT_NOTICE);
			exit(0);
		default:
			fprintf(stderr, "%s: unknown option code 0x%x\n", program, c);
		case '?':
			usage();
		}
	if(debug_flag)
		fprintf(stderr, "%s: interface=%s timeout=%d num=%d size=%d\n",
			program, iface, timeout, retries, pack_size);
	if(optind != argc - 1) {
		fprintf(stderr, "%s: missing target address\n", program);
		usage();
	}
	arg = argv[optind];
}

void sig_alarm(int signo)
{
}

static void set_sighandlers()
{
	struct sigaction sa;

	sa.sa_handler = sig_alarm;
	sa.sa_flags = 0;
	sigfillset(&sa.sa_mask);	/* Block spurious signals during handler */
	sigdelset(&sa.sa_mask, SIGINT);	/* But normal signals should kill us */
	sigdelset(&sa.sa_mask, SIGQUIT);
	sigdelset(&sa.sa_mask, SIGTERM);
	if(sigaction(SIGALRM, &sa, NULL) < 0) {
		perror("sigaction");
		exit(1);
	}
}

static int linkloop(int sock, const uint8_t mac_src[], const uint8_t mac_dst[])
{
	int ret;
	struct llc_packet spack;
	struct llc_packet rpack;
	int datasize = pack_size - sizeof(struct llc) - sizeof(struct ether_header);

	mk_test_packet(&spack, mac_src, mac_dst, pack_size, 0);
	if(alarm(timeout) < 0) {
		perror("alarm");
		exit(1);
	}
	send_packet(sock, iface, &spack);
	total_sent++;
	ret = recv_packet(sock, &rpack);
	if(ret == 0) {		/* timeout */
		fprintf(stderr, "  ** TIMEOUT (%d seconds)\n", timeout);
		total_timeout++;
		return 0;
	}
	if(spack.eth_hdr.ether_type != rpack.eth_hdr.ether_type) {
		total_bad++;
		printf("  ** BAD RECEIVED LENGTH: sent = %d, received = %d\n",
			spack.eth_hdr.ether_type,
			rpack.eth_hdr.ether_type);
		return 0;
	}
	if(memcmp(spack.eth_hdr.ether_dhost, rpack.eth_hdr.ether_shost, IFHWADDRLEN) != 0) {
		total_bad++;
		printf("  ** ROGUE RESPONDER: received from %s\n",
			mac2str(rpack.eth_hdr.ether_shost));
		return 0;
	}
	if(memcmp(spack.data, rpack.data, datasize) != 0) {
		total_bad++;
		printf("  ** BAD RESPONSE\n");
		dump_packet(&rpack);
		return 0;
	}
	total_good++;
	return 1;
}

int main(int argc, char * const argv[])
{
	int sock;
	int i;

	uint8_t mac_src[IFHWADDRLEN];
	uint8_t mac_dst[IFHWADDRLEN];

	handle_options(argc, argv);

	if(!parse_address(mac_dst, arg)) {
		fprintf(stderr, "%s: bad DST address - %s\n", program, arg);
		return 1;
	}

	printf("Link connectivity to LAN station: %s (HW addr %s)\n", arg, mac2str(mac_dst));

	/* Open a socket */
	if((sock = socket(PF_PACKET, SOCK_PACKET, htons(ETH_P_802_2))) == -1) {
		perror("socket");
		return 1;
	}
	if(debug_flag)
		fprintf(stderr, "Getting MAC address of interface '%s'\n", iface);
	get_hwaddr(sock, iface, mac_src);
	if(debug_flag)
		fprintf(stderr, "Testing via %s (HW addr %s)\n", iface, mac2str(mac_src));

	set_sighandlers();
	for(i = 0; i < retries; i++) {
		if(linkloop(sock, mac_src, mac_dst))
			;
		if(debug_flag)
			printf("Retry %d...\n", i);
	}

	if(total_sent == total_good)
		printf("  -- OK -- %d packets\n", total_sent);
	if(total_good + total_bad == 0)
		printf("  -- NO RESPONSE --\n");
	return EXIT_SUCCESS;
}
